{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3 Analysis Step #1; Quality Control\n",
    "\n",
    "### 3.1 Loading the patient data into Galaxy\n",
    "We need to load the patient data you are going to work on during this project. All patient data is allready on the server and you first need to load it into your Galaxy <strong>History</strong> to work with. To do this fist go to the <strong>Shared Data</strong> menu and click on <strong>Data Libraries</strong>\n",
    "\n",
    "<img src=\"pics/dataLibrariesMenu.png\"><br />\n",
    "\n",
    "<img src=\"pics/patientDataMenu.png\"><br />\n",
    "\n",
    "Open the `patientData` library. Here you will find the sequence data from five patients, called samples 1 to 5. As you can see each sample consists of two files. This is because the forward and reverse reads are split up into two files (*paired-end sequencing*). The files with **R1** in the name are the forward reads and the reverse reads have **R2** in their name, remember this as some tools require to select the proper file. \n",
    "\n",
    "Each group will be assigned a patient. Please contact the teacher to hear which patient is assigned to your group. Once you know which patient you have you can select the reads. If for example `patient1` is your patient, you can select the reads from `Sample1`. Check the boxes for the R1 and R2 of Sample1 and click the <strong>to History</strong> button. Select the `Unnamed` history and click on Import. \n",
    "\n",
    "Note that the filetype is in the `.fastq.gz` format which means the files are compressed using `gzip`. Since FastQ data is in basic text format this compresses very well to reduce file size. Here Galaxy shows its use by *unzipping* the data before running a tool if a tool doesn't allow working on compressed data (most do). Furthermore, the data type is defined as `fastqsanger` which is important because some manufacturers calculate the quality score differently. Without the proper data type a tool such as `FastQC` (see below) doesn't know when a read is of sufficient quality.\n",
    "\n",
    "<img src=\"pics/patientData.png\"><br />\n",
    "\n",
    "After you have loaded the sequence data in your history it is available to work on. You can go back by clicking on the `Analyze Data` link. If all went correct you should now see the two sequence files in your History of Galaxy. \n",
    "\n",
    "<img src=\"pics/historyPatient.png\"><br />\n",
    "\n",
    "#### Note for FireFox users:\n",
    "\n",
    "At some point you might see an error page when *viewing* the contents of a file in Galaxy (by clicking on the `eye`-icon in the history. This error can be fixed by changing one of the security settings of your browser. You can do this by typing <strong>about:config</strong> in the url bar. In the options you are presented with, search for: <strong>network.http.accept-encoding</strong>. Right click and select <strong>Modify</strong>. Remove the text that is there and change it to <strong>*</strong>. If you don't see this error there is no need for changing this setting.\n",
    "\n",
    "Every time a dataset is loaded into the *history* panel (from a manual action, like we just did or from the output of a tool), you can have a quick look by clicking on the name of the dataset. In this case it will show you the first read (and a bit..) of the loaded library. A couple of actions can be taken on each item in your History such as *View Data, Edit Attributes* and *Delete*. When you click on the name of an item in the History additional actions such as downloading the data, viewing the Galaxy (job) background information or visualising the data appear. Go ahead poke the `eye`-button now and have a look at the raw data.\n",
    "\n",
    "### 3.2 Check the quality of the patient data\n",
    "When you start to analyse NGS data it is very important to get a feeling for the data. What does the data look like? What is the quality of my data? You do not (although you *can*) want to draw conclusions on low quality data. For some common errors the data can be corrected. You would first need to identify what is wrong with the data. To look at many (quality) aspects of our imported datasets, we are going to start a tool named [**FastQC**](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).\n",
    "In the <strong>Tools</strong> panel, open the <strong>Diagnostic Genome Analysis</strong> category and click on the <strong>FastQC</strong> tool. The settings of the FastQC tool should appear in the middle section. A help is ussualy shown when you open a tool as well as the scientific literature associated with the tool for further references. For the FastQC tool we can load short read data from our history. Also a contaminant list can be uploaded (for example primers from a pre-process step can sometimes end up in the data and give all sort of problems downstream). In the submodule part of the FastQC setting you can specify wich subparts of the tools need te be run. For now we are not going to bother you with settings other than the short read data. \n",
    "\n",
    "Select <strong>Multiple Datasets</strong> under the <strong>Short read data from your current history</strong> and press <strong>Execute</strong>.<br />\n",
    "<img src=\"pics/fastQCSettings.png\">\n",
    "<br />\n",
    "\n",
    "When you clicked Execute, the jobs were started and added to the History. Each item in the history has a number which increments with one for each new item. When the jobs are finished they appear in green. Four new items have appeared in the History. If you look closely, you can see that FastQC has been run on data 1 and 2 from the History. FastQC has generated two types of reports: <strong>RawData</strong> and a <strong>Webpage</strong>. We will have a look at the Webpages for both short read datasets. Click on the **View Data** (eye-icon) for the webpage output for data 1 <br /><img src=\"pics/fastQCHistory.png\"><br />\n",
    "\n",
    "In the middle panel the webpage with the results for data 1 should open. A summary of the quality checks performed is visable and gives you a quick overview of the checks that need your attention.\n",
    "\n",
    "#### Assignment 2 read quality\n",
    "\n",
    "Report on the total number of sequences analysed, the number of sequences flagged as poor quality, the sequence length and GC percentage. Next look at the <strong>per base sequence quality</strong>, in the plot the sequence position is plotted on the x-axis and the base quality is plotted on the y-axis. Each bar in the plot represents the distribution over all sequences analysed in your dataset at this position. Also look at the Webpage from data 2 and compare the plots. Do you see any differences between the plots and would you use this dataset for further analysis? \n",
    "\n",
    "For further information on the box plot, please have a look at: https://en.wikipedia.org/wiki/Box_plot\n",
    "<br /><img src=\"pics/fastQCOPlot.png\"><br />\n",
    "\n",
    "\n",
    "### 3.3 Cleaning the Data\n",
    "\n",
    "Now that we took a first look at the data and its quality, we will modify the data to make sure that we get data of the highest quality to perform our analysis on. Most likely the current data contains (many) reads of low quality or which are too short for performing the next steps. In the following two chapters we will finalize our look at the data quality by performing *read trimming* (section 3.3 below) to remove these reads (or parts of reads) with low quality. Once we have a data set of acceptable quality we will perform the next and most important analysis step; *read mapping* (chapter 4), then we visualize the resulting mapping (assignment 5, manually inspecting the data) and perform a *pileup* operation (chapter 5).\n",
    "\n",
    "As you can see in the FastQC report, the read quality drops at the end of the reads (this is normal for the Illumina protocol and has to do with the specific sequencing technology used!). Also the spread in quality at some positions may be greater than you would like to see. In the next step we are going to remove reads that are of low quality or are just too short to be used. To make the DNA available to be sequenced, primers have been annealed to the DNA. These primers should also be removed.\n",
    "\n",
    "From the <strong>Tools menu </strong>select the **[Trimmomatic](http://bioinformatics.oxfordjournals.org/content/26/14/1783)** tool. \n",
    "* Check if the *paired end* data is selected.\n",
    "* Select *Pair of Datasets* as the *Input Type* (default) and select your patient R1 and R2 files. \n",
    "* Perform an *initial IlluminaCLIP* step and select the *TruSeq3 (paired-ended, for miSeq and HiSeq)* adapters. \n",
    "* From the *Trimmomatic operation* select the Sliding window trimming and choose 4 bases to average across, the quality should be 20 (as a minimum). \n",
    "* Click on the plus sign (*Insert Trimmomatic Operation* button) and\n",
    "    * Select the *Drop reads below a specified length*\n",
    "    * Set the minimum read length to **70** (can you think of a reason why you do not want to have too many short reads?).\n",
    "* Execute the tool.\n",
    "<img src=\"pics/trimmomatic.png\">\n",
    "\n",
    "Trimmomatic will give **4** new files as output, **read** the help of the Trimmomatic tool (can be found below the tool settings in Galaxy) to understand what each file contains. For next steps in our analysis it is very important that the forward and reverse reads are in the same order in the files. If for any reason one of the forward or reverse reads was removed from one of the files, the two files will *not be in order* any more. For this reason Trimmomatic will remove both reads from the files. If one read was below the threshold and was removed, it's paired partner will be written to an unpaired output file (if it was above the threshold). In the output you will find 4 files: one for the forward paired and one for the reverse paired (these are still in order) and one for the unpaired forward and one for the unpaired reverse. Rename the two files that you will use for the next step to something like *`Trimmed Reads [ID] R1 00X`* etc.\n",
    "\n",
    "\n",
    "#### Assignment 3 read filtering\n",
    "\n",
    "* Report on the **number of reads** that where removed for each dataset.\n",
    "    * Create new FastQC plots of the *Per Base Sequence Quality* (on the cleaned, paired data).\n",
    "    * Compare the plots with the original uncleaned data and report on the differences.\n",
    "* Perform multiple FASTQ cleaning runs (each time on the original data). \n",
    "    * Change a setting and report on the number of reads removed. \n",
    "    * Create and save the *Per Base Sequence Quality* plots for each run.\n",
    "    * **Report** (in a table for instance) what the effect is of each setting in the Trimmomatic tool.\n",
    "    * Pick the best (combination of) settings and copy the R1 and R2 trimmed datasets to a **new** history in Galaxy, this data will be used for the next analysis step.\n",
    "    * Please remove any unused output of Trimmomatic from the previous history, as this migth take up a few gigabytes of space. If you delete these files (they can also be 'undeleted') they are just marked for actual deletion after a few months."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
