{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7 Analysis Step #5; Annovar\n",
    "\n",
    "<!-- http://nar.oxfordjournals.org/content/38/16/e164 -->\n",
    "\n",
    "In the - for now - final Galaxy workflow step for analyzing our patient, we are going to annotate all our filtered variants to determine the severity of the variations we determined last time with the `Varscan` tool. This involves comparing all our variants against a number of databases containing *annotation data* for known variants. We will use our filtered `VCF`-file resulting from Varscan so we need to upload our newly created VCF file (the output from `deliverable5.py`) to Galaxy. Make sure that Galaxy recognizes the file type by selecting the correct file type when uploading or change it afterwards using the *Edit attributes* button.\n",
    "\n",
    "<!--\n",
    "<strong>NOTE:</strong> In the previous step we removed the first 24 lines from the vcf file. These first 24 lines are actualy needed by this step. So before we proceed, we should rerun our previuous script and not delete those lines, but write them also to the output file. You can upload your filtered vcf file to galaxy to work on. Select the <strong>upload button</strong> which is in the upper left corner next to the <strong>Tools</strong>. Select the <strong>Choose local file</strong>. Set the type to <strong>vcf</strong> and the genome to <strong>hg19[Human feb. 2009(GRCh37/hg19)]</strong>\n",
    "-->\n",
    "<img src=\"pics/upload.png\">\n",
    "\n",
    "## 7.1 Running Annovar\n",
    "\n",
    "Select the [**ANNOVAR**](http://annovar.openbioinformatics.org/en/latest/) tool by searching for its name or selecting it from the *Diagnostic Genome Analysis* tool menu. This tool offers a lot of configuration options in the form of selectable sources for comparison. Note that not all of these sources are actually available in our Galaxy server, so take care when selecting the proposed options below.\n",
    "\n",
    "Set the following settings and data sources from the tool configuration as follows:\n",
    "* **Reference**: `hg19` [Human Feb. 2009 (GRCh37/hg19)]\n",
    "* **Select file to annotatte**: your uploaded **filtered** VCF file\n",
    "* **Select filetype**: `VCF4 file`\n",
    "* **Select Gene Annotations**: `Select all`\n",
    "* **Select dbSNP version(s) to annotate with**: `138 (hg19 only)`\n",
    "* **Select 1000Genomes Annotation**: `2015aug (hg19/hg38) (6 populations: AMR,AFR,EUR,EAS,SAS,ALL)`\n",
    "* **Select Exome Variant Server version(s) to annotate with**: `ESP6500si European Americans`\n",
    "* **Annotate with ExAC 03?**: `Yes`\n",
    "* **CLINVAR Annotation? (hg19 only)**: `Yes`\n",
    "* **Select functional impact scores (LJB2)**: `Select all`\n",
    "* **Also get predictions where possible?**: `Yes`\n",
    "\n",
    "With all these options selected, Execute the tool.\n",
    "\n",
    "<img src=\"pics/annovar.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download the resulting `.tabular` file to your computer and open it with Excel or the Linux alternative (you might need to change the file extension to `.tsv` (for *tab separated values* format) for Excel to recognize the file). As you will see, the file looks like the input-VCF file, but a number of columns are added containing the annotation data from all the selected data sources. \n",
    "* Create a table in your report where you summarize the added columns (very!) briefly;\n",
    "    * The database where the data comes from\n",
    "    * The result or value (is it an identifier, a percentage, etc.). See the links in 7.3 below.\n",
    "\n",
    "## 7.2 Documentation; Writing an Introduction\n",
    "\n",
    "As we've now run the final step in our analysis pipeline we *should* know everything about the process going from raw sequencing reads (the FASTQ file contents) to annotated variants. If *any* of the previous steps is not yet clear you should revisit those step(s) and carefully re-read the documentation.\n",
    "\n",
    "With our current knowledge you should be able to write a complete **Introduction** chapter, please see the instructions and the report template available on [Blackboard](https://blackboard.hanze.nl/bbcswebdav/pid-4379773-dt-content-rid-31189649_2/xid-31189649_2). Create a new 'Introduction' document in your (shared) OneNote folder and start writing this chapter.\n",
    "\n",
    "The Introduction chapter is also a common place to reference to books, web-resources and publications which is why you are also asked to write a **References** chapter (in a new document). Please carefully read the instructions in the template report as this contains examples for all three mentioned resources. Also note that most Galaxy tools are published and they contain a *how-to-cite* link (or even an export button) which you can use to compile your list.\n",
    "\n",
    "## 7.3 Programming Assignment; Analyzing Variant Annotation\n",
    "\n",
    "Given the possibly thousands of annotated variants with data from multiple data sources, it's time to do one last *filtering* on the data to come up with a list of variants of which we can say, with some certainty, that they are related to the condition. Whether we want to keep a variant or not depends on the annotation added by ANNOVAR and it is your task to come up with an approach for **combining** data from these data sources to *rank* the annotated variants (explained further in section 7.4). To do this, we need to get the relevant information from the ANNOVAR for each variant.\n",
    "\n",
    "**Notes**: \n",
    "* for some of these data sources listed below, only a few of your variants are actually annotated\n",
    "* the list below links to the Annovar documentation regarding this value; read these documents to at least *know* what they mean.\n",
    "* this is the **minimal** list of properties to store\n",
    "\n",
    "From all the annotation columns we need to get the following data for each variant:\n",
    "* The gene name from the `RefSeq_Gene` column,\n",
    "* the gene 'function' from the `RefSeq_Func` column (see the [table of possible values](http://annovar.openbioinformatics.org/en/latest/user-guide/gene/#output-file-1-refseq-gene-annotation) for this column),\n",
    "* the [`dbsnp`](http://annovar.openbioinformatics.org/en/latest/articles/dbSNP/) identifier,\n",
    "* the 1000-genomes data for the `EUR` population group (the others are not of interest),\n",
    "* the [`sift`](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#-sift-annotation) value from the `LJB2_SIFT` column,\n",
    "* the [`polyphen`](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#-polyphen-2-annotation) value from one of the `LJB2_PolyPhen2_*` columns and\n",
    "* the [`clinvar`](http://annovar.openbioinformatics.org/en/latest/user-guide/filter/#clinvar-annotations) value.\n",
    "\n",
    "To make proper use of this annotation data we will once again write a *filter* Python program that *parses* the ANNOVAR output file and filters out these values. Instead of writing this to a new file, we will store the results into a MySQL database consisting three linked tables.\n",
    "\n",
    "<!-- MINOR EDIT\n",
    "of three tables as shown below:\n",
    "\n",
    "<img src=\"pics/Tablet.png\">\n",
    "<img src=\"pics/anno-db.png\">\n",
    "-->\n",
    "\n",
    "This programming assignment consists of four deliverables (6 through 9) where the first one focuses on parsing the ANNOVAR output file (and simply prints a number of values per variant). Deliverable 7 is for creating a *regular-expression pattern* for parsing the gene name from the `RefSeq_Gene` column. Deliverable 8 asks to create a database schema for storing all relevant information and deliverable 9 is for inserting the parsed data into the this database.\n",
    "\n",
    "### Deliverable 6; ANNOVAR file parsing\n",
    "\n",
    "Given the list of required fields from the ANNOVAR output file, write a program that reads in an ANNOVAR output file (you can use the following [example data (`example.txt`)](https://bitbucket.org/mkempenaar/diagnosticgenomeanalysis/raw/master/templates/data/example.txt) to 'extract' this data. Note that there is no template available for this assignment.\n",
    "\n",
    "Notes:\n",
    "* Pre-define the columns to extract, i.e. `columns = [10, 11, 15:18, 20:24]`\n",
    "* Combine this with the parsed *header* string (`\"chromosome\\tbegin\\tend\\treference\\tobserved\\tCHROM\\tPOS...\"`)\n",
    "* Output of this deliverable should be a **printed** summary of the data for **each** variant, purely used to 'prove' that the program is working:\n",
    "    * A single line giving a variant *position* overview, i.e. gene name, chromosome, reference base, etc.\n",
    "    * A set of **indented** lines, with one of the above annotation values per line (i.e. `\"\\tdbsnp: rs1621733\"`)\n",
    "    * A blank line before the next variant\n",
    "\n",
    "\n",
    "<!--based on the [Deliverable 6 Template Python Script](https://bitbucket.org/mkempenaar/diagnosticgenomeanalysis/raw/master/templates/deliverable6.py) Nope... -->\n",
    "\n",
    "### Deliverable 7; Gene-name parsing\n",
    "\n",
    "The `RefSeq_Gene` column in the ANNOVAR output file contains a somewhat complex representation of the gene name that the variant is associated with. It could contain just the text '`ERBB4`' if the variant lies on an exon of that gene, but it could also be '`BIN1(dist=32600),CYP27C1(dist=43909)`' if the variant lies in the intergenic region *between* the `BIN1` and `CYP27C1` genes where the `dist=***` denotes the distance to these genes. This assignment filteres out the gene names, and thus removes these distance numbers.\n",
    "\n",
    "Create a regular expression pattern that catches all gene names from the sometimes complex value of the `RefSeq_Gene` column. The [Deliverable 7 Template Python Script](https://bitbucket.org/mkempenaar/diagnosticgenomeanalysis/raw/master/templates/deliverable7.py) contains a function that gets the value from this annotation column and returns the actual gene name(s) by 'extracting' it using a regular expression and filtering it using standard Python code.\n",
    "\n",
    "Example data has been made available at the online regular expression test website [regex101](https://regex101.com/r/rmxDs5/2/) and click on the **fork regex** button to start designing your own regular expression. The `Test String` field contains seven lines taken from the `RefSeq_Gene` column that you can use to test your pattern on. You are free to add data to this field from your own ANNOVAR output file.\n",
    "\n",
    "The following block shows the example input data and the expected output data. As you can see, we are only interested in retaining the gene name and not the `NONE`, `LOC****` and `LIN****` fields. Please read the instructions in the template script carefully for further details and don't expect to easily complete this assignment using *just* regular expressions (combine them with simple `if/else` statements too."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "'TNNI3(NM_000363:exon5:c.371+2T>A)'                      ----> 'TNNI3'\n",
    "'TSHZ3(dist=65732),THEG5(dist=173173)'                   ----> 'TSHZ3/THEG5'\n",
    "'ACTR3BP2(dist=138949),NONE(dist=NONE)'                  ----> 'ACTR3BP2'\n",
    "'BIN1(dist=32600),CYP27C1(dist=43909)'                   ----> 'BIN1/CYP27C1'\n",
    "'LOC101927282(dist=1978702),LINC01519(dist=14658)'       ----> '-'\n",
    "'NBPF10,NBPF20'                                          ----> 'NBPF10/NBPF20'\n",
    "'ERBB4'                                                  ----> 'ERBB4'\n",
    "'LOC100507291'                                           ----> '-'\n",
    "'NONE'                                                   ----> '-'\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Deliverable 8; Database Schema\n",
    "\n",
    "Create an SQL database schema that consists of three tables to store information from the (*i*) Chromosomes, (*ii*) Genes and (*iii*) Variants. These tables should have a relation making it possible to i.e. get all variants for a specific gene (by name) or count the number of variants per chromosome etc.\n",
    "\n",
    "Carefully study the output of the Annovar tool to decide where you can find the information and to define the datatype to use for storage.\n",
    "\n",
    "For this assignment you are asked to create both a graphical representation of your database (image, place this in your notebook) and an SQL file (`deliverable8.sql`) containing the `CREATE TABLE` statements (add this file to your forked repository). Your SQL file should begin with one or more [`DROP TABLE ... IF EXISTS`](https://dev.mysql.com/doc/refman/5.7/en/drop-table.html) statements.\n",
    "\n",
    "### Deliverable 9; Database Interaction\n",
    "\n",
    "With the ANNOVAR file parsed and the gene names set correclty it is time to create and fill the database. Instead of creating an SQL file that you can import into the database, we will write Python code to directly connect to MySQL and execute the `CREATE` and `INSERT` statements. The supporting library to use is [`mysql.connector`](https://dev.mysql.com/doc/connector-python/en/connector-python-introduction.html) available for Python 2 and 3 (side note; to view available Python libraries on the/your system, use `pip3 freeze`). The [official documentation](https://dev.mysql.com/doc/connector-python/en/connector-python-introduction.html) contains all the information and [examples](https://dev.mysql.com/doc/connector-python/en/connector-python-examples.html) that you need to create a fully functional database interacting Python program. \n",
    "\n",
    "Your program should have the following functionality:\n",
    "\n",
    "* Accept the following input parameters using the `argparse` Python library:\n",
    "    + Database `name`\n",
    "    + Database `username`\n",
    "    + Database `password`\n",
    "    + ANNOVAR annotation file\n",
    "    + (optional) database schema file (`.sql` file for (re-)creating the database)\n",
    "* **Test** the given database name and credentials (i.e. perform a simple [`SELECT 1`](https://stackoverflow.com/questions/3668506/efficient-sql-test-query-or-validation-query-that-will-work-across-all-or-most) query)\n",
    "    * The program thus needs to connect to the database (on the `mysql.bin` MySQL server)\n",
    "* **Test** if the given ANNOVAR file exists and if it contains tab-separated data (i.e., the length of the tab-splitted header should be > 1)\n",
    "* If given **and** the previous test(s) pass, execute the `.sql` file (this will `DROP` all existing table(s))\n",
    "* Parse the ANNOVAR file (use the code from deliverable 9)\n",
    "    * Either build up a proper object containing the data and logic for parsing (see 'bonus' below) or\n",
    "    * Process the parsed data on-the-fly;\n",
    "        * Once you have the chromosome information; insert into the `Chromosome` table,\n",
    "        * Once you have the gene information (parsed by your deliverable 7 code), insert into the `Gene` table,\n",
    "        * Once you have the variant information, insert into the `Variant` table.\n",
    "* Once all data has been parsed and inserted into the database, a small report should be printed that states the number of rows in each table. This should be added to your OneNote journal file.\n",
    "    * And the database and/ or file connections should be closed\n",
    "\n",
    "The program should also properly report any problems encoutered.\n",
    "\n",
    "* Bonus points available for:\n",
    "    * Proper use of OO-techniques\n",
    "    * Dividing coherent functionality accross multiple modules\n",
    "\n",
    "As you can see this program contains many aspects and experience tells us that *just starting to program* results in messy code that requires a lot of [refactoring](https://en.wikipedia.org/wiki/Code_refactoring) to reach the required quality. Therefore, the first task is to create a **deliverable 9 template** file that does **not** contain any function code, but only function [*stubs*](https://en.wikipedia.org/wiki/Method_stub) (define the function **name**, its **arguments** and a **return value** or a simple `pass` statement. Please keep in mind some of the programming quality standards and best practises for Python, amongst:\n",
    "* Function names contain no uppercase letters (only a Class starts with an uppercase)\n",
    "* Try to think of very clear function/ method names (important!)\n",
    "* Add a tripple-quoted docstring to each method and explain its functionality\n",
    "* Try to use a maximum of two or three arguments for a function/ method\n",
    "\n",
    "Once you have your template finished and saved as `deliverable9.py`, run the `pylint deliverable9.py` command to see if you missed any coding standards (try to aim at a pylint 'grade' of >7). Once this template is completed, add it to your repository (change where necessary):\n",
    "\n",
    "```\n",
    "git add templates/deliverable9.py\n",
    "git commit -m \"Added a template for deliverable 9\"\n",
    "git push -u origin master\n",
    "```\n",
    "\n",
    "<span style='color:#800000; font-weight:bold; '>This template needs to be checked before you start developing the actual program! </span>\n",
    "\n",
    "Once verified you can start *filling in your functions*!\n",
    "\n",
    "\n",
    "## 7.4 Finding Variants of Interest\n",
    "\n",
    "Given the knowledge we have on the data stored in the database, we can now formulate questions to this data to find a set of variants which are most likely linked to the cardiomyopathy condition. This condition cannot be identified by a single variant but is often caused by several variants and the combination not only determines *if* someone suffers from cardiomyopathy but also how *badly* their condition is. \n",
    "\n",
    "The physisian who diagnoses patients uses the genetic information of the found variants as one of the sources for confirming the condition. However, presenting a list of 2000+ variant positions with a list of numbers from some tools is not how that works in practice; we need to present a short, ordered list of variants that are found to be involved. \n",
    "\n",
    "First, give a short overview of what we have in our database. After inserting the data you showed an overview of the number of rows per table, now, create some queries to report on:\n",
    "* the number of variants per chromosome and\n",
    "* the number of variants per gene\n",
    "\n",
    "Next, define your own sorting, filtering, selecting, etc. procedure to filter and order a top-list (a maximum of 50, but less is fine too) of variants that you'd present as aid to a diagnoses. For instance, you might have a variant that had a hit in the Clinvar database which directly links it to cardiomyopathy, this should probably be somewhere around the top of your list. Again, look at the output description that you've written earlier to come up with combination(s) of values. There is no golden standard for selecting these variants so it is very important to properly document and argument your decisions. A tip might be to first sort on SIFT score, then on the PolyPhen scores (sort them correctly!) and filter for *exonic* variants (intronic are almost always less significant). \n",
    "\n",
    "As an end product for this step you are **required** to show a nice table in your report that comes from a **single** query. This table should contain all relevant information for each of your selected variants, i.e. the chromomsome, position, reference and variant nucleotide, gene name and all relevant scores. This means that you will jave to `JOIN` the tables together for this final reporting step.\n",
    "\n",
    "Take the top-10 of these variants and try to (briefly!) answer the following questions for each variant:\n",
    "* Can you find a link between cardiomyopathy and the variant?\n",
    "* What is the effect of the variant?\n",
    "* Is it validated?\n",
    "* Are there scientific studies supporting the evidence?\n",
    "* What does the score indicate?\n",
    "* What else can you find about the variant?\n",
    "\n",
    "You can use the following online sources amongst others:  \n",
    "* http://www.ncbi.nlm.nih.gov/snp/  \n",
    "* http://www.ncbi.nlm.nih.gov/variation/tools/1000genomes/  \n",
    "* http://www.ncbi.nlm.nih.gov/projects/SNP/  \n",
    "\n",
    "## 7.5 Galaxy Workflow\n",
    "\n",
    "Now that we have completed our analysis steps we can create a Galaxy workflow as a last exercise. In this workflow we will combine all the tools that we have used in Galaxy - including our custom VCF-filter Python program - so that when we receive new patient data in the future we can analyze that dataset with a single click!\n",
    "\n",
    "There are many [online resources](https://galaxyproject.org/learn/advanced-workflow/) giving instructions and examples for creating Galaxy workflows, use these resources to create a workflow including all used analysis tools. Note that you can convert Galaxy histories into workflow(s), which saves a lot of time with Annovar for instance as this tool requires a lot of configuration.\n",
    "\n",
    "Once you have a workflow, make sure it is shared with your project partner and create a screenshot of it for your report (try to fit all tools in a single image).\n",
    "\n",
    "Note; this assignment has a very low priority and should be done once you are satisfied with all other parts of your work! At the bare minimum, open your last history (including Annovar) and explore/ play with the `Extract Workflow` option.\n",
    "\n",
    "## Project Deadlines\n",
    "\n",
    "There are multiple deadlines for finalizing this project:\n",
    "* Programming assignments: all graded deliverables (5, 6, 7, 8 & 9) *must* be available on your BitBucket repository on **Friday November the 1st, 24.00h** (even if incomplete, make sure that the latest version(s) are available).\n",
    "* Journal/'report': the OneNote documents are downloaded for grading on **Thursday November the 14th** and the final grade is available on Friday the 15th of November. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
